Simplify __split_range

EricGustin opened the pull request:
This PR simplifies the `__Class.__split_range` static method.

`__Class.__split_range` should throw an error if `pattern` is not of valid structure. A check for this should probably be done before returning, but I have omitted this part at this point in time as I have not read enough of the codebase to know if this check is already being done somewhere else, so I am waiting for your feedback on that.

The previous implementation didn't return a tuple when `count == 1` or `count ==2` (it would return a list of length 2), going against the function header. In addition, this method really just wants the first and last character in the `pattern` string, so there is no need to break it up into 3 cases. Also, the previous implementation would return `tuple("-", "-")` even in the case of invalid input such as `pattern="foobar"`.

manoss96 commented:
Hey there! The change that you propose would actually break the code. You can confirm this yourself by running the tests. However, you've brought up some good points.

> __Class.__split_range should throw an error if pattern is not of valid structure.

Under normal circumstances, this should be the case as this would not only protect from having "wrong" patterns fed to this function, but it would also indicate what the correct pattern actually is, which by the way is `\\?.-\\?.`, where "." includes newlines. This means that the current header note stating that characters within the pattern should not be escaped is actually false and it must be removed. However, given that this function is private and not to be invoked by anyone other than the class's own methods, I think it's OK to skip this pattern check. Besides, having it would actually bring the test coverage below 100% as I can't really test feeding this function a "wrong" pattern.

> 
The previous implementation didn't return a tuple when count == 1 or count ==2 (it would return a list of length 2), going against the function header.

You're right. In fact, the code would break (though not always) if this function were to return a tuple, as tuples can't have their values re-assigned, which is something the code might do. Therefo, both the return type hint as well as the third if/else case must be switched from a tuple to a list.

I'll leave this PR open for a few days in case you are interested in implementing said changes. If you do, please include a new note indicating the expected pattern structure, i.e. `\\?.-\\?.`

EricGustin commented:
I see, the expected pattern was different than what I had assumed & gathered from the method note about not having any escaped characters. I updated the function return type, and the docstring, and removed the need for assigning the `split` and `rsplit` functions to a variable. I'm looking forward to contributing more in the future.

manoss96 commented:
There seems to be a bug within the code. Try returning `["-", "-"]` instead of `list("-", "-")` within "else".
